### PR TITLE
Fix #710 ServiceTracker TSAN warnings

### LINKED ISSUE #710: listenerToken data race in ServiceTrackerPrivate
The ServiceTrackerPrivate holds a listener token. In ServiceTracker::Close(), the listener token is passed through a std::move() outside of the synchronized region for the ServiceTrackerPrivate, creating a data race when Open() and Close() calls are made concurrently.

This race is warned by TSan, but is currently suppressed.

Moving the RemoveListener call in ServiceTracker::Close() will likely resolve the race, but the possibility of re-entrancy should be examined before implementing this solution.

**To Reproduce:**
- Remove ServiceTracker lines from tsan_suppressions.txt
- Run usFrameworkTests (specifically ServiceTrackerTest.ServiceTrackerConcurrentOpenClose) in TSan

### PR DESCRIPTION
**Pull request by tcormackMW**
Fixing TSAN warnings in `ServiceTracker.hpp`. Some were valid, specifically the race when accessing the variable `d->listenerToken`, but others were TSAN bugs related to waiting on condition variables, and so those were silenced

### CODE REVIEW

**Review by jeffdiclemente — APPROVED** [review]

File: framework/include/cppmicroservices/detail/ServiceTracker.hpp [hunk 1]
```diff
@@ -139,8 +139,16 @@ void ServiceTracker<S, T>::Open()
 template<class S, class T>
 void ServiceTracker<S, T>::Close()
 {
+  // Copy required to prevent data race in call to context.RemoveListener()
+  ListenerToken tokenCpy;
+  {
+    auto l = d->Lock();
+    US_UNUSED(l);
+    std::swap(d->listenerToken, tokenCpy);
```
Comment (carneyweb) [inline framework/include/cppmicroservices/detail/ServiceTracker.hpp:147]: The comment and this line of code don't match up. std::swap does exactly that, swaps the state of the two objects, effectively doing this:

    auto tmp = a;
    a = b;
    b = tmp;

So the state of d->listenerToken above is reset to the default state for the ListenerToken class.

**Review by carneyweb — CHANGES_REQUESTED** [review]

**Review by carneyweb — APPROVED** [review]
Fix it, then ship it.

File: framework/include/cppmicroservices/detail/ServiceTracker.hpp (hunk 1 above)
Comment (jeffdiclemente) [reply framework/include/cppmicroservices/detail/ServiceTracker.hpp:147]: good catch. yes, the comment is not telling the truth

File: framework/include/cppmicroservices/detail/ServiceTracker.hpp [hunk 2]
```diff
@@ -151,9 +151,18 @@ namespace cppmicroservices
     void
     ServiceTracker<S, T>::Close()
     {
+
+        /* Swap default ListenerToken with d-listenerToken required to prevent data race in call to
```
Comment (jeffdiclemente) [inline framework/include/cppmicroservices/detail/ServiceTracker.hpp:155]: nit-pick: typo in `d-listenerToken`

The comment is a little confusing, specifically the use of the word `default`. Did you mean `default constructored`?  Also, this comment doesn't explain the "why", which is more imporatnt.

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

**Review by jeffdiclemente — APPROVED** [review]

**Pull request merged**